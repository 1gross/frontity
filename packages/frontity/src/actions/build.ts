export default () => {
  console.log("`frontity build` is currently under development");
};
