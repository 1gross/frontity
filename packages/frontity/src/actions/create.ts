import { resolve } from "path";
import ora from "ora";
import { prompt, Question } from "inquirer";
import create from "../functions/create";
import errorLogger from "../utils/error";
import { EventEmitter } from "events";
import { Options } from "../functions/create/types";

export default async (name: string, { typescript, useCwd }) => {
  const options: Options = {};

  if (!name) {
    const questions: Question[] = [
      {
        name: "name",
        type: "input",
        message: "Enter a name for the project:",
        default: "my-frontity-project"
      },
      {
        name: "theme",
        type: "input",
        message: "Enter a starter theme to clone:",
        default: "@frontity/mars-theme"
      },
      {
        name: "packages",
        type: "input",
        message: "Enter a list of Frontity packages to install:",
        default: "frontity, @frontity/file-settings",
        filter: (input: string) => input.split(/[\s,]+/)
      }
    ];

    const answers = await prompt(questions);
    options.name = answers.name;
    options.packages = answers.packages;
    options.theme = answers.theme;
    console.log();
  } else {
    options.name = name;
  }

  options.typescript = typescript;
  options.path = useCwd ? process.cwd() : resolve(process.cwd(), options.name);
  options.emitter = new EventEmitter();

  options.emitter.on("create", (message, action) => {
    if (action) ora.promise(action, message);
    else console.log(message);
  });

  options.emitter.on("error", error => {
    errorLogger(error);
  });

  await create(options);
};
