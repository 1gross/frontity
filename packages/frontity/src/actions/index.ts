export { default as create } from "./create";
export { default as createPackage } from "./create-package";
export { default as dev } from "./dev";
export { default as build } from "./build";
export { default as serve } from "./serve";
