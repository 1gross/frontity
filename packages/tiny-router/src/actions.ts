import TinyRouter from "..";
import { parsePath, addFinalSlash } from "./utils";

let isPopState = false;

export const set: TinyRouter["actions"]["router"]["set"] = ({
  state,
  libraries
}) => routeOrParams => {
  const { getParams, getRoute } = libraries.source;
  const params = getParams(routeOrParams);
  const route = getRoute(routeOrParams);

  state.router.path = params.path;
  state.router.page = params.page;
  state.router.query = params.query;

  if (state.frontity.platform === "client" && !isPopState) {
    window.history.pushState(params, "", route);
  } else {
    isPopState = false;
  }
};

export const init: TinyRouter["actions"]["router"]["init"] = ({
  state,
  actions,
  libraries
}) => {
  if (state.frontity.platform === "server") {
    // Populate the router info with the initial path and page.
    const params = libraries.source.getParams(state.frontity.initial);
    state.router.path = params.path;
    state.router.page = params.page;
    state.router.query = params.query;
  } else {
    // Replace the current url with the same one but with state.
    const { path, page, query } = state.router;
    window.history.replaceState({ path, page, query }, "");
    // Listen to changes in history.
    window.addEventListener("popstate", ({ state: params }) => {
      isPopState = true;
      actions.router.set(params);
    });
  }
};
