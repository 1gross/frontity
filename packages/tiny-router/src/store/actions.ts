import { State, PathOrObj } from "../../type";

const set = (state: State) => (pathOrObj: PathOrObj) => {
  const path = typeof pathOrObj === "string" ? pathOrObj : pathOrObj.path;
  const page = typeof pathOrObj === "string" ? null : pathOrObj.page;

  state.router.path = path;
  state.router.page = page;

  if (typeof window !== "undefined") {
    window.history.pushState({ path, page }, "", path);
  }
};

const init = (state: State) => {
  if (typeof window !== "undefined") {
    window.addEventListener("popstate", ({ state: { path, page } }) =>
      set(state)({ path, page })
    );
  }
};

export default { set, init };
