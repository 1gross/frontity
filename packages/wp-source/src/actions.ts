import WpSource from "../";
import { parse, normalize, concatPath } from "./libraries/route-utils";
import { wpOrg, wpCom } from "./libraries/patterns";
import { getMatch } from "./libraries/get-match";
import { redirect } from "./libraries/redirect";

const actions: WpSource["actions"]["source"] = {
  fetch: ({ state, libraries }) => async link => {
    const { source } = state;

    const { handlers, redirections } = libraries.source;

    // Get route and route params
    const route = normalize(link);
    const routeParams = parse(link);

    // Get current data object
    const data = source.data[route];

    // return if the data that it's about to be fetched already exists
    if (data) return;

    // init data
    source.data[route] = {
      isReady: false,
      isFetching: true
    };

    // get and execute the corresponding handler based on path
    try {
      // transform path if there is some redirection
      let { path } = routeParams;
      path = redirect(path, redirections);

      // get the handler for this path
      const handler = getMatch(path, handlers);
      if (!handler) throw new Error(`No matching handler for route "${route}"`);

      await handler.func({ route, params: handler.params, state, libraries });
      // everything OK
      source.data[route] = {
        ...source.data[route],
        isFetching: false,
        isReady: true
      };
    } catch (e) {
      console.log(e);
      // an error happened
      source.data[route] = {
        is404: true,
        isFetching: false,
        isReady: false
      };
    }
  },

  init: ({ state, libraries }) => {
    const { api, isWpCom } = state.source;

    libraries.source.api.init({ api, isWpCom });

    // handlers & redirections:
    const { handlers, redirections } = libraries.source;

    const patterns = isWpCom ? wpCom : wpOrg;
    handlers.push(...patterns);

    const {
      subdirectory,
      homepage,
      postsPage,
      categoryBase,
      tagBase
    } = state.source;

    if (subdirectory) {
      const pattern = concatPath(subdirectory, "/:subpath*");
      redirections.push({
        name: "subdirectory",
        priority: 10,
        pattern,
        func: ({ subpath }) => (subpath ? `/${subpath}/` : "/")
      });
      redirections.push({
        name: "ignore root",
        priority: 10,
        pattern: "/(.*)",
        func: () => ""
      });
    }

    if (homepage) {
      const pattern = "/";
      redirections.push({
        name: "homepage",
        priority: 20,
        pattern,
        func: () => concatPath(homepage)
      });
    }

    if (postsPage) {
      const pattern = concatPath(postsPage);
      redirections.push({
        name: "posts page",
        priority: 20,
        pattern,
        func: () => "/"
      });
    }

    if (categoryBase) {
      // add new direction
      const pattern = concatPath(categoryBase, "/:subpath+");
      redirections.push({
        name: "category base",
        priority: 20,
        pattern,
        func: ({ subpath }) => `/category/${subpath}/`
      });
      // remove old direction
      redirections.push({
        name: "ignore default category base",
        priority: 20,
        pattern: concatPath("/category/(.*)/"),
        func: () => ""
      });
    }

    if (tagBase) {
      // add new direction
      const pattern = concatPath(tagBase, "/:subpath+");
      redirections.push({
        name: "tag base",
        priority: 20,
        pattern,
        func: ({ subpath }) => `/tag/${subpath}/`
      });
      // remove old direction
      redirections.push({
        name: "ignore default tag base",
        priority: 20,
        pattern: concatPath("/tag/(.*)/"),
        func: () => ""
      });
    }
  }
};
export default actions;
