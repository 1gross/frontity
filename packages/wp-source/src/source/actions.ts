import WpSource from "../../";
import { normalizePath } from "./utils";
import { wpOrg, wpCom } from "../libraries/patterns/";

const actions = ({
  libraries
}: {
  libraries: WpSource["libraries"];
}): WpSource["actions"]["source"] => ({
  fetch: state => async pathOrObj => {
    const { source } = state;
    const { resolver } = libraries.source;

    let path = typeof pathOrObj === "string" ? pathOrObj : pathOrObj.path;
    let page = typeof pathOrObj === "string" ? 1 : pathOrObj.page;
    if (!page) page = 1;

    // transform whole links to paths
    path = normalizePath(path);

    // Get current data object
    const data = source.dataMap[path];

    // return if the data that it's about to be fetched already exists
    if (data && data.isArchive && data.pages[page]) return;

    // init data if needed
    if (!data) source.dataMap[path] = {};

    // get and execute the corresponding handler based on path
    try {
      const { handler, params } = resolver.match(path);
      await handler(source, { path, params, page, libraries });
      source.dataMap[path].isReady = true;
    } catch (e) {
      console.warn(`An error ocurred fetching '${path}:'\n`, e);
      source.dataMap[path].is404 = true;
    }

    // end fetching
    source.dataMap[path].isFetching = false;
  },

  init: state => {
    const { apiUrl, isCom } = state.source;
    const { api, resolver } = libraries.source;

    api.init({ apiUrl, isCom });

    const patterns = isCom ? wpCom : wpOrg;
    patterns.forEach(({ pattern, handler }) => resolver.add(pattern, handler));
  }
});

export default actions;
