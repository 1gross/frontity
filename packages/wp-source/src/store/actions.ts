import { Action } from "../types";
import { normalizePath } from "./helpers";

type PathOrObj =
  | string
  | {
      path: string;
      page?: number;
    };

export const fetch: Action<PathOrObj> = async (ctx, pathOrObj) => {
  const state = ctx.state.source;
  const effects = ctx.effects.source;

  let { path, page = 1 } =
    typeof pathOrObj === "object" ? pathOrObj : { path: pathOrObj };

  // transform whole links to paths
  path = normalizePath(path);

  // Get current data object
  const data = state.dataMap[path];

  // return if the data that it's about to be fetched already exists
  if (data && page && data.isArchive && data.pages[page]) return;

  // init data if needed
  if (!data) state.dataMap[path] = {};

  // get and execute the corresponding handler based on path
  try {
    const { handler, params } = effects.resolver.match(path);
    await handler(ctx, { path, params, page });
    state.dataMap[path].isReady = true;
  } catch (e) {
    console.warn(`An error ocurred fetching '${path}:'\n`, e);
    state.dataMap[path].is404 = true;
  }

  // end fetching
  state.dataMap[path].isFetching = false;
};
