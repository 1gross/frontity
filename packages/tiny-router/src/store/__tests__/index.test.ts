import actions from "../actions";
// import { createStore } from "@frontity/connect";
// import state from "../state";

// const config = {
//   state: {
//     router: state,
//     frontity: {
//       url: "https://example.com"
//     },
//   },
//   actions: {
//     router: actions,
//   }
// };

// let store: any;
// beforeEach(() => {
//   store = createStore(config);
// });

// describe("state", () => {
//   test.todo("addtests");
//   test("url without page", () => {
//     expect(store.state.router.url).toBe("https://example.com/");
//   });
//   test("url with page = 1", () => {
//     expect(store.state.router.url).toBe("https://example.com/");
//   });
//   test("url with page > 1", () => {
//     expect(store.state.router.url).toBe("https://example.com/");
//   });
// });

const state = {
  router: {
    path: "",
    page: null,
    url: "https://example.com",
  },
  frontity: {
    url: "https://example.com",
  }
}

describe("actions set", () => {
  // test("set", async () => {
  //   store.actions.router.set("/some-post");
  //   expect(store.state.router.url).toBe("https://example.com/some-post");
  // });
  test("set() should work just with a path", () => {
    actions.set(state)("/some-post");
    expect(state).toMatchSnapshot();
  });
  test("set() should work with path and page", () => {
    actions.set(state)({ path: "/category/some-category", page: 2 });
    expect(state).toMatchSnapshot();
  });
  test.todo("init() should add event listener to handle popstate events");
});
