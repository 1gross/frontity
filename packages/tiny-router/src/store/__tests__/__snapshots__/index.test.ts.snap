// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions set set() should work just with a path 1`] = `
Object {
  "frontity": Object {
    "url": "https://example.com",
  },
  "router": Object {
    "page": null,
    "path": "/some-post",
    "url": "https://example.com",
  },
}
`;

exports[`actions set set() should work with path and page 1`] = `
Object {
  "frontity": Object {
    "url": "https://example.com",
  },
  "router": Object {
    "page": 2,
    "path": "/category/some-category",
    "url": "https://example.com",
  },
}
`;
